src/dmqproto/node/neo/request/Consume.d

```diff
@@ -1004,3 +1004,470 @@ public abstract scope class ConsumeProtocol_v2
+                    size_t record_length = (*this.value_buffer).length;
+                    (*this.record_batch) ~= (&record_length)[0 .. 1];
+                    (*this.record_batch) ~= *this.value_buffer;
+
+                    if ((*this.record_batch).length >= min_batch_length)
+                    {
+                        this.sendBatch();
+                        if (!this.waitForFeedback())
+                            break SendAndWaitLoop; // Received a `Stop` message.
+                    }
+                }
+
+                // All records have been popped from the storage. Wait for
+                //  - a record to be pushed in the storage or
+                //  - a `Flush` signal (if there are records in the batch) or
+                //  - a `Stop` message from the client.
+                this.outer.resume_fiber_on_push = true;
+                this.outer.resume_fiber_on_flush = !!(*this.record_batch).length;
+
+                auto sigcode = this.outer.request_event_dispatcher.nextEvent(
+                    this.fiber,
+                    Signal(SignalCode.Pushed), Signal(SignalCode.Flush),
+                    Signal(SignalCode.Stop)
+                ).signal.code;
+
+                this.outer.resume_fiber_on_push = false;
+                this.outer.resume_fiber_on_flush = false;
+
+                switch (sigcode)
+                {
+                    case SignalCode.Pushed:
+                        // Continue `SendAndWaitLoop` to pop and send the record
+                        break;
+
+                    case SignalCode.Flush:
+                        // Send the imcomplete batch and wait for a feedback
+                        // i.e. `Continue` or `Stop` message from the node.
+                        this.sendBatch();
+                        if (this.waitForFeedback())
+                            break; // Received a `Continue` message.
+                        else
+                            break SendAndWaitLoop; // Received a `Stop` message.
+
+                    case SignalCode.Stop:
+                        break SendAndWaitLoop;
+
+                    default:
+                        assert(false);
+                }
+            }
+
+            // The client signalled `Stop`. Send the remaining batch, if it
+            // contains records, acknowledge with `Stopped` and exit.
+
+            if ((*this.record_batch).length)
+                this.sendBatch();
+
+            this.outer.request_event_dispatcher.send(this.fiber,
+                ( ed.Payload payload )
+                {
+                    payload.addConstant(MessageType_v3.Stopped);
+                }
+            );
+
+            // Tell the `MessageReceiver` to quit.
+            this.outer.request_event_dispatcher.signal(
+                this.outer.ed, SignalCode.Stopped
+            );
+        }
+
+        /// Sends the current batch of records to the client and clears it.
+        private void sendBatch ( )
```

Comment:
As `sendBatch` and `waitForFeedback` are always called together, it might make sense to combine them into one. (It's actually invalid to call one without the other, right?)

Reply:
They are not always called together: When the request ends, in `RecordSender.fiberMethod`` after the `SendAndWaitLoop`, the pending batch is sent without waiting for a feedback.

Reply:
Ah yes, I see.